File: computer-networks/assg-1/python/tcp-client.py
# File: tcp-client.py

# Author: Anindya Kundu <anindya.k22@outlook.com>
# Date: 2020-06-02

# Reference: https://docs.python.org/3/library/
# Reference: https://stackoverflow.com/a/17668009

"""
A Simple TCP client program.

This client can chat with one client at a time. This client initiates
a conversation, to which the server replies. The server and client
get alternate changes to send a message. The conversation ends when
one of them sends a "bye".

Functions:

- `_chat()`: conducts the chat with the client
- `_recv_msg()`: receives data in small chunks and assembles message
- `_send_msg()`: sends message with the length as header

How To Use This Module
======================
1. run server script using [python3 tcp-server.py]
2. run this script using [python3 tcp-client.py]
"""


import socket
import sys
from os import system


def _recv_msg(connection):
    """
    Assembles message by receiving data in small chunks.

    Parameters `connection`: client socket object

    Return message assembled by concatenating all received data chunks
    """

    SIZE = 16       # size of data chunk in bytes
    message = ""    # message received

    sendlen = int(connection.recv(8).decode())      # message length
    recvlen = 0

    while not recvlen == sendlen:
        # receive data in small chunks and concatenate
        data = connection.recv(SIZE).decode()
        recvlen += len(data)
        message += data

    return message


def _send_msg(connection, message):
    """
    Sends message length and message.

    Parameters:

    - `connection`: client socket object
    - `message`: message string to send

    Return void
    """

    # pad message length string with 0s to get 8 bytes
    msglen = str(len(message))
    zeros = 8 - min(len(msglen), 8)
    for _ in range(zeros):
        msglen = "0" + msglen

    connection.sendall(msglen.encode())     # send message length
    connection.sendall(message.encode())    # send message


def _chat(connection):
    """
    Waits for the client to initiate the conversation. Prints the
    client's message and takes a message to send to the client.
    This process continues until one of then says "bye".

    Parameters:

    - `connection`: server socket object
    """

    # infinite loop with one turn to each of client and server per iteration
    while True:
        message = input('CLIENT: ')
        _send_msg(connection, message)
        if message == 'bye':
            break

        response = _recv_msg(connection)
        print('SERVER: %s' % response, file=sys.stderr)
        if response == 'bye':
            break


if __name__ == "__main__":
    """
    Attribute is __main__ when module is run as main program.
    """

    # clear console
    system('clear')

    # create TCP/IP socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    # connect socket to the server's listening port
    server_address = ('localhost', 4200)
    print('connecting to %s port %s\n' % server_address, file=sys.stderr)
    sock.connect(server_address)

    try:
        # call chat function to conduct chat
        _chat(sock)

    finally:
        # clean up connection
        print('\nclosing socket', file=sys.stderr)
        sock.close()

File: computer-networks/assg-1/c/tcp-client.c
/**
 * @file Simple TCP client program
 * @brief This client can chat with one server. This client initiates
 * a conversation, to which the server replies. The server and client
 * get alternate changes to send a message. The conversation ends when
 * one of them sends a "bye".
 *
 * @ref https://pubs.opengroup.org/onlinepubs/7908799/
 * @ref https://www.man7.org/linux/man-pages/man7/
 *
 * ****************************************************************************
 * compile using [cc -o client tcp-client.c]
 * run using [./client]
 * ****************************************************************************
 *
 * @author Anindya Kundu <anindya.k22@outlook.com>
 * @date 2020-06-02
 */

#include <stdio.h>      /** @include standard buffered input/output */
#include <sys/types.h>  /** @include data types */
#include <sys/socket.h> /** @include main sockets header */
#include <netinet/in.h> /** @include Internet address family */
#include <arpa/inet.h>  /** @include definitions for internet operations */
#include <string.h>     /** @include string operations */
#include <stdlib.h>     /** @include standard library definitions */
#include <unistd.h>     /** @include standard symbolic constants and types */

#define PORT 4200            /** @def PORT number of Server */
#define SADR struct sockaddr /** @def short reference symbol */
#define BUFF 1024            /** @def maximum buffer size in bytes */

/** chat function declaration */
void _chat(int);

/**
 * Main function.
 *
 * @param argc - command line arguments' count
 * @param argv - command line arguments
 * @returns exit status
 */
int main(int argc, char *argv[])
{
    /** clear terminal (stdlib.h) */
    system("clear");    // unix-like
    // system("cls");      // cmd/powershell

    /** server socket file descriptor */
    int sockfd;
    /** server address descriptor */
    struct sockaddr_in ser_addr;

    // create socket and verify
    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) == -1)
    {
        perror("-- socket creation failed, error");
        exit(1);
    }
    else
    {
        printf("-- socket created successfully\n");
    }

    // initialize all bytes in ser_addr to 0
    bzero(&ser_addr, sizeof(ser_addr));

    // assign IP, PORT
    ser_addr.sin_family = AF_INET;
    ser_addr.sin_port = htons(PORT);
    ser_addr.sin_addr.s_addr = inet_addr("127.0.0.1");

    // connect client socket to server socket
    if (connect(sockfd, (SADR *)&ser_addr, sizeof(ser_addr)) == -1)
    {
        perror("-- connection with server failed, error");
        exit(1);
    }
    else
    {
        printf("-- connected to server\n");
    }

    // call function for chat
    _chat(sockfd);

    // close socket after chatting
    close(sockfd);
    printf("\n-- socket closed\n");
}

/**
 * Conducts the chat between client and server.
 *
 * @private
 * @param connfd - client socket file descriptor
 */
void _chat(int sockfd)
{
    char buff[BUFF];

    printf("\n");
    // infinite loop for chat
    while (1)
    {
        // clear buffer and store client message
        bzero(buff, sizeof(buff));
        printf("CLIENT: ");
        scanf("%s", buff);

        // send buffer to server
        write(sockfd, buff, sizeof(buff));

        // if message is "bye", exit chat
        if (strncmp("bye", buff, 3) == 0)
        {
            printf("\n-- client exit\n");
            break;
        }

        // clear buffer and receive server message
        bzero(buff, sizeof(buff));
        read(sockfd, buff, sizeof(buff));
        printf("SERVER: %s\n", buff);

        // if message is "bye", exit chat
        if (strncmp("bye", buff, 3) == 0)
        {
            printf("\n-- client exit\n");
            break;
        }
    }
}
